### PR TITLE
PT-1312 add fields to alarm objects

### PR DESCRIPTION
**Pull request by franck-fourel**
- adding AlarmLevel, AlarmCode and AlarmLabel
- remaining fields will be added later after #52

### CODE REVIEW

File: data/types/device/alarm/alarm.go [hunk 1]
```diff
@@ -85,6 +120,15 @@ func (a *Alarm) Validate(validator structure.Validator) {
...
 	}
+	validator.String("ackStatus", a.AckStatus).OneOf(AckStatuses()...)
+	if a.EventID != nil {
+		validator.String("alarmLevel", a.AlarmLevel).Exists().OneOf(AlarmLevels()...)
+		validator.String("alarmCode", a.AlarmCode).Exists()
+		validator.String("alarmLabel", a.AlarmLabel).Exists()
+		validator.String("ackStatus", a.AckStatus).Exists()
+		timeValidator := validator.String("updateTime", a.UpdateTime)
+		timeValidator.AsTime(types.TimeFormat).Exists()
+	}
```
Comment (CedricPerotto) [inline data/types/device/alarm/alarm.go]: From the spec :
> fields are mandatory if `alarmTYpe` is set to `handset`. They remain optional for any other values.

The logic here is that fields are validated when eventId exists. And updateTime (for thr time format) and alarmLevel(for the list of accepted values) are validated only if enventId is set.

File: docs/UpdatedData.md [hunk 2]
```diff
@@ -223,28 +223,25 @@ In the below example, the physical activity is entered on the handset at 8:00am.
...
-- `alarmLabel`: label or description of the alarm. This field is optional. 
-- `eventId`: unique Id of the event generated by the client system. This ID will be used to reconciliate data for the same event. 
-- `eventType`: `start | stop` is the type of event for the given alarm.
-  - `start`: alarm created by the system
-  - `stop`: the system has received the patient acknowledge. 
-
-For a given alarm that has been acknowledged by the patient, we will receive 2 deviceEvents of subType `alarm` with the same `eventId`:
-- the first one gives the creation time on the system, `eventType`: `start`
-- the second one gives the patient acknowledge, `eventType`: `stop`
+The below fields are mandatory if `alarmTYpe` is set to `handset`. They remain optional for any other values.
```
Comment (CedricPerotto) [inline docs/UpdatedData.md]:
```suggestion
The below fields are mandatory if `alarmType` is set to `handset`. They remain optional for any other values.
```

File: docs/UpdatedData.md [hunk 3]
```diff
@@ -223,28 +223,25 @@ In the below example, the physical activity is entered on the handset at 8:00am.
...
   "type": "deviceEvent",
   "subType": "alarm",
   "alarmType": "handset",
+  "eventId": "123456789",
   "alarmLevel": "alarm", 
   "alarmCode": "123456",
   "alarmLabel": "Label of the alarm",
-  "eventId": "123456789",
-  "eventType": "alarm",
+  "status": "acknowledged",
```
Comment (CedricPerotto) [inline docs/UpdatedData.md]:
```suggestion
  "ackStatus": "acknowledged",
```

**Review by CedricPerotto — CHANGES_REQUESTED** [review]

File: data/types/device/alarm/alarm.go (hunk 1 above)
Comment (franck-fourel) [reply data/types/device/alarm/alarm.go]: reworked based on alarmType.

File: data/types/device/alarm/alarm.go [hunk 4]
```diff
@@ -120,14 +123,23 @@ func (a *Alarm) Validate(validator structure.Validator) {
...
+			timeValidator := validator.String("updateTime", a.UpdateTime)
+			timeValidator.AsTime(types.TimeFormat).Exists()
+		} else {
+			validator.String("ackStatus", a.AckStatus).OneOf(AckStatuses()...)
+			validator.String("alarmCode", a.AlarmCode)
+			validator.String("alarmLabel", a.AlarmLabel)
+			validator.String("ackStatus", a.AckStatus)
+			timeValidator := validator.String("updateTime", a.UpdateTime)
+			timeValidator.AsTime(types.TimeFormat)
+		}
```
Comment (CedricPerotto) [inline data/types/device/alarm/alarm.go]: alarmLevel is not always validated with the list of Alarm Levels
```suggestion

	alarmLevelValidator := validator.String("alarmLevel", a.AlarmLevel)
	alarmLevelValidator.OneOf(AlarmLevels()...)
	
	ackStatusValidator := validator.String("ackStatus", a.AckStatus)
	ackStatusValidator.OneOf(AckStatuses()...)
	
	timeValidator := validator.String("updateTime", a.UpdateTime)
	timeValidator.AsTime(types.TimeFormat)
	
	if a.AlarmType !=nil && *a.AlarmType == AlarmTypeHandset {
		alarmLevelValidator.Exists()
		ackStatusValidator.Exists()
		validator.String("eventID", a.EventID).Exists()
		validator.String("alarmCode", a.AlarmCode).Exists()
		validator.String("alarmLabel", a.AlarmLabel).Exists()
		timeValidator.Exists()
	}
```

**Review by CedricPerotto — CHANGES_REQUESTED** [review]

**Review by CedricPerotto — APPROVED** [review]

**Comment by franck-fourel** [comment]
@yvon-dblg if you want to have a look ?

File: data/types/device/alarm/alarm_test.go [hunk 5]
```diff
@@ -64,6 +77,9 @@ func CloneAlarm(datum *alarm.Alarm) *alarm.Alarm {
 		}
 	}
 	clone.StatusID = pointer.CloneString(datum.StatusID)
+	clone.AlarmLevel = pointer.CloneString(datum.AlarmLevel)
+	clone.AlarmCode = pointer.CloneString(datum.AlarmCode)
+	clone.AlarmLabel = pointer.CloneString(datum.AlarmLabel)
```
Comment (yvon-dblg) [inline data/types/device/alarm/alarm_test.go:83]:
```suggestion
	clone.AlarmLabel = pointer.CloneString(datum.AlarmLabel)
	clone.AckStatus = pointer.CloneString(datum.AckStatus)
	clone.UpdateTime = pointer.CloneString(datum.UpdateTime)
```
Missing AckStatus & UpdateTime

File: docs/UpdatedData.md [hunk 6]
```diff
@@ -223,28 +223,25 @@ In the below example, the physical activity is entered on the handset at 8:00am.
...
-For a given alarm that has been acknowledged by the patient, we will receive 2 deviceEvents of subType `alarm` with the same `eventId`:
-- the first one gives the creation time on the system, `eventType`: `start`
-- the second one gives the patient acknowledge, `eventType`: `stop`
+The below fields are mandatory if `alarmType` is set to `handset`. They remain optional for any other values.
+- `eventId`: unique Id of the event generated by the client system. This ID will be used to reconciliate data for the same event.
+- `alarmLevel`: `alarm | alert`.  
+- `alarmCode`: code of the alarm.  
+- `alarmLabel`: label or description of the alarm.  
+- `ackStatus`: this fields gives the acknowledge status of the alarm that can take one of the following values `new | acknowledged | outdated`. 
+- `updateTime`: this timestamp gives the last time the alarm was updated. . It takes the same format as `time` field.
```
Comment (yvon-dblg) [inline docs/UpdatedData.md]:
```suggestion
- `updateTime`: this timestamp gives the last time the alarm was updated. It takes the same format as `time` field.
```

**Review by yvon-dblg — COMMENTED** [review]
Few remarks / typo.
Not sure for this, should we set an upper limit in size for the "free" fields?

File: data/types/device/alarm/alarm_test.go (hunk 5 above)
Comment (franck-fourel) [reply data/types/device/alarm/alarm_test.go:83]: good catch
It shows as well that handset was not tested correctly for Normalize step

**Comment by franck-fourel** [comment]
> Few remarks / typo.
> Not sure for this, should we set an upper limit in size for the "free" fields?

what would you propose? something like a length that would be less than 100 chars

**Comment by yvon-dblg** [comment]
> 
> 
> > Few remarks / typo.
> > Not sure for this, should we set an upper limit in size for the "free" fields?
> 
> what would you propose? something like a length that would be less than 100 chars

I'm not sure how much text we need. If possible stick to a exp 2 limit (128, 256 ?)

**Comment by franck-fourel** [comment]
@yvon-dblg up to you to decide if it's ok based on your last comment and the latest changes

**Review by yvon-dblg — APPROVED** [review]
LGTM

**Pull request merged**